tgui/packages/tgui/interfaces/pda/pda_messenger.js

```diff
@@ -0,0 +1,262 @@
+        <LabeledList.Item label="Messenger Functions">
+          <Button
+            icon="trash"
+            color="bad"
+            onClick={() => act("Clear", { option: "Convo" })}>
+            Delete Conversations
+          </Button>
+        </LabeledList.Item>
+      </LabeledList>
+      {body}
+    </Box>
+  );
+};
+
+const MessengerList = (props, context) => {
+  const { act, data } = useBackend(context);
+
+  const {
+    convopdas,
+    pdas,
+    charges,
+    silent,
+    toff,
+  } = data;
+
+  return (
+    <Box>
+      <LabeledList>
+        <LabeledList.Item label="Messenger Functions">
+          <Button
+            selected={!silent}
+            icon={silent ? "volume-mute" : "volume-up"}
+            onClick={() => act("Toggle Ringer")}>
+            Ringer: {silent ? "Off" : "On"}
+          </Button>
+          <Button
+            color={toff ? "bad" : "green"}
+            icon="power-off"
+            onClick={() => act("Toggle Messenger")}>
+            Messenger: {toff ? "Off" : "On"}
+          </Button>
+          <Button
+            icon="bell"
+            onClick={() => act("Ringtone")}>
+            Set Ringtone
+          </Button>
+          <Button
+            icon="trash"
+            color="bad"
+            onClick={() => act("Clear", { option: "All" })}>
+            Delete All Conversations
+          </Button>
+        </LabeledList.Item>
+      </LabeledList>
+      {!toff && (
+        <Box>
+          {!!charges && (
+            <LabeledList>
+              <LabeledList.Item label="Cartridge Special Function">
+                {charges} charges left.
+              </LabeledList.Item>
+            </LabeledList>
+          )}
+          {!convopdas.length && !pdas.length && (
+            <Box>
+              No current conversations
+            </Box>
+          ) || (
+            <Box>
+              <PDAList title="Current Conversations"
+                pdas={convopdas}
+                msgAct="Select Conversation" />
+              <PDAList title="Other PDAs" pdas={pdas} msgAct="Message" />
+            </Box>
+          )}
+        </Box>
+      ) || (
+        <Box color="bad" mt={2}>
+          Messenger Offline.
+        </Box>
+      )}
+    </Box>
+  );
+};
+
+const PDAList = (props, context) => {
+  const { act, data } = useBackend(context);
+
+  const {
+    pdas,
+    title,
+    msgAct,
+  } = props;
+
+  const {
+    charges,
+    plugins,
+  } = data;
+
+  if (!pdas || !pdas.length) {
+    return (
+      <Section level={2} title={title}>
+        No PDAs found.
+      </Section>
+    );
+  }
+
+  return (
+    <Section level={2} title={title}>
+      {pdas.map(pda => (
+        <Box key={pda.Reference}>
+          <Button
+            icon="arrow-circle-down"
+            content={pda.Name}
+            onClick={() => act(msgAct, { target: pda.Reference })} />
+          {!!charges && plugins.map(plugin => (
+            <Button
+              key={plugin.ref}
```

Comment:
Sidenote: I thought we always try to use .UID() instead of refs for things?

Reply:
I usually do but messenger always uses `ref` for everything and I didnt want to change everything to use `UID` en masse, but I will if its majorly required.